UTF16: reading file is fine, but writing file is garbage

### Summary

When reading from UTF16 file, helix decode it correctly, but when writing back the file on disk, it write back garbage data.

### Reproduction Steps

<!-- Ideally provide a key sequence and/or asciinema.org recording. --> 

#### I tried this:

1. `wget https://raw.githubusercontent.com/stain/encoding-test-files/master/utf16.txt` - download a sample UTF16 file inside this project
1. `file utf16.txt` - print `utf16.txt: Unicode text, UTF-16, little-endian text`
1. `hx`
1. In helix, type `:x`
1. `file utf16.txt` - now print `utf16.txt: ISO-8859 text`  :warning:  

#### I expected this to happen:

helix to convert the file to UTF8 - or keep UTF16 little endian

#### Instead, this happened:

helix written something else completely - ASCII character survive the change of encoding but any character outside ASCII is not Unicode characters.


#### File content 

<details><summary>hexdump before file edit</summary>

```
00000000  ff fe 70 00 72 00 65 00  6d 00 69 00 e8 00 72 00  |..p.r.e.m.i...r.|
00000010  65 00 20 00 69 00 73 00  20 00 66 00 69 00 72 00  |e. .i.s. .f.i.r.|
00000020  73 00 74 00 0a 00 70 00  72 00 65 00 6d 00 69 00  |s.t...p.r.e.m.i.|
00000030  65 00 00 03 72 00 65 00  20 00 69 00 73 00 20 00  |e...r.e. .i.s. .|
00000040  73 00 6c 00 69 00 67 00  68 00 74 00 6c 00 79 00  |s.l.i.g.h.t.l.y.|
00000050  20 00 64 00 69 00 66 00  66 00 65 00 72 00 65 00  | .d.i.f.f.e.r.e.|
00000060  6e 00 74 00 0a 00 1a 04  38 04 40 04 38 04 3b 04  |n.t.....8.@.8.;.|
00000070  3b 04 38 04 46 04 30 04  20 00 69 00 73 00 20 00  |;.8.F.0. .i.s. .|
00000080  43 00 79 00 72 00 69 00  6c 00 6c 00 69 00 63 00  |C.y.r.i.l.l.i.c.|
00000090  0a 00 01 d8 00 dc 20 00  61 00 6d 00 20 00 44 00  |...... .a.m. .D.|
000000a0  65 00 73 00 65 00 72 00  65 00 74 00 0a 00        |e.s.e.r.e.t...|
000000ae
```

</details>

<details><summary>hexdump after file edit</summary>

```
00000000  70 72 65 6d 69 e8 72 65  20 69 73 20 66 69 72 73  |premi.re is firs|
00000010  74 0a 70 72 65 6d 69 65  26 23 37 36 38 3b 72 65  |t.premie&#768;re|
00000020  20 69 73 20 73 6c 69 67  68 74 6c 79 20 64 69 66  | is slightly dif|
00000030  66 65 72 65 6e 74 0a 26  23 31 30 35 30 3b 26 23  |ferent.&#1050;&#|
00000040  31 30 38 30 3b 26 23 31  30 38 38 3b 26 23 31 30  |1080;&#1088;&#10|
00000050  38 30 3b 26 23 31 30 38  33 3b 26 23 31 30 38 33  |80;&#1083;&#1083|
00000060  3b 26 23 31 30 38 30 3b  26 23 31 30 39 34 3b 26  |;&#1080;&#1094;&|
00000070  23 31 30 37 32 3b 20 69  73 20 43 79 72 69 6c 6c  |#1072; is Cyrill|
00000080  69 63 0a 26 23 36 36 35  36 30 3b 20 61 6d 20 44  |ic.&#66560; am D|
00000090  65 73 65 72 65 74 0a                              |eseret.|
00000097
```

</details>


### Helix log

<details><summary>~/.cache/helix/helix.log</summary>

```
2023-04-01T21:43:04.411 helix_view::clipboard [DEBUG] Using wl-copy+wl-paste to interact with the system and selection (primary) clipboard
2023-04-01T21:43:04.412 helix_vcs [ERROR] Error {
    context: "failed to open git repo",
    source: Discover(
        NoGitRepositoryWithinFs {
            path: "/tmp",
            limit: "/",
        },
    ),
}
2023-04-01T21:43:04.412 helix_vcs [ERROR] failed to open diff base for /tmp/utf16.txt
2023-04-01T21:43:04.412 helix_vcs [ERROR] Error {
    context: "failed to open git repo",
    source: Discover(
        NoGitRepositoryWithinFs {
            path: "/tmp",
            limit: "/",
        },
    ),
}
2023-04-01T21:43:04.412 helix_vcs [ERROR] failed to obtain current head name for /tmp/utf16.txt
2023-04-01T21:43:04.412 helix_view::editor [DEBUG] editor status: Loaded 1 file.
2023-04-01T21:43:04.415 helix_tui::backend::crossterm [DEBUG] The keyboard enhancement protocol is supported in this terminal (checked in 3.104306ms)
2023-04-01T21:43:04.415 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:04.812 helix_term::application [DEBUG] received editor event: IdleTimer
2023-04-01T21:43:04.903 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:05.062 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:05.215 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:05.367 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:05.767 helix_term::application [DEBUG] received editor event: IdleTimer
2023-04-01T21:43:05.975 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:06.150 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:06.376 helix_term::application [DEBUG] received editor event: IdleTimer
2023-04-01T21:43:06.382 helix_view::document [DEBUG] submitting save of doc 'Some("/tmp/utf16.txt")'
2023-04-01T21:43:06.382 helix_term::job [DEBUG] waiting on jobs...
2023-04-01T21:43:06.383 helix_view::document [DEBUG] doc 1 revision updated 0 -> 0
2023-04-01T21:43:06.383 helix_term::commands::typed [DEBUG] quitting...
2023-04-01T21:43:06.383 helix_view::document [DEBUG] id 1 modified - last saved: 0, current: 0
2023-04-01T21:43:06.383 helix_term::job [DEBUG] waiting on jobs...
2023-04-01T21:43:06.383 helix_term::job [DEBUG] waiting on jobs...
```

</details>


### Platform

Linux - Fedora 38

### Terminal Emulator

kitty 0.26.5

### Helix Version

helix 23.03

--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -12,7 +12,7 @@ pub use typed::*;
 use helix_core::{
     char_idx_at_visual_offset, comment,
     doc_formatter::TextFormat,
-    find_first_non_whitespace_char, find_workspace, graphemes,
+    encoding, find_first_non_whitespace_char, find_workspace, graphemes,
     history::UndoKind,
     increment, indent,
     indent::IndentStyle,
@@ -32,7 +32,7 @@ use helix_core::{
 };
 use helix_view::{
     clipboard::ClipboardType,
-    document::{EncodingBom, FormatterError, Mode, SCRATCH_BUFFER_NAME},
+    document::{FormatterError, Mode, SCRATCH_BUFFER_NAME},
     editor::{Action, Motion},
     info::Info,
     input::KeyEvent,
@@ -5084,7 +5084,8 @@ async fn shell_impl_async(
     let output = if let Some(mut stdin) = process.stdin.take() {
         let input_task = tokio::spawn(async move {
             if let Some(input) = input {
-                helix_view::document::to_writer(&mut stdin, EncodingBom::default(), &input).await?;
+                helix_view::document::to_writer(&mut stdin, (encoding::UTF_8, false), &input)
+                    .await?;
             }
             Ok::<_, anyhow::Error>(())
         });

--- a/helix-term/src/commands.rs
+++ b/helix-term/src/commands.rs
@@ -12,7 +12,7 @@ pub use typed::*;
 use helix_core::{
     char_idx_at_visual_offset, comment,
     doc_formatter::TextFormat,
-    encoding, find_first_non_whitespace_char, find_workspace, graphemes,
+    find_first_non_whitespace_char, find_workspace, graphemes,
     history::UndoKind,
     increment, indent,
     indent::IndentStyle,
@@ -32,7 +32,7 @@ use helix_core::{
 };
 use helix_view::{
     clipboard::ClipboardType,
-    document::{FormatterError, Mode, SCRATCH_BUFFER_NAME},
+    document::{EncodingBom, FormatterError, Mode, SCRATCH_BUFFER_NAME},
     editor::{Action, Motion},
     info::Info,
     input::KeyEvent,
@@ -5084,7 +5084,7 @@ async fn shell_impl_async(
     let output = if let Some(mut stdin) = process.stdin.take() {
         let input_task = tokio::spawn(async move {
             if let Some(input) = input {
-                helix_view::document::to_writer(&mut stdin, encoding::UTF_8, &input).await?;
+                helix_view::document::to_writer(&mut stdin, EncodingBom::default(), &input).await?;
             }
             Ok::<_, anyhow::Error>(())
         });

--- a/helix-term/tests/test/commands/write.rs
+++ b/helix-term/tests/test/commands/write.rs
@@ -407,3 +407,41 @@ async fn test_write_fail_new_path() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+#[tokio::test(flavor = "multi_thread")]
+async fn test_write_utf_bom_file() -> anyhow::Result<()> {
+    // "ABC" with utf8 bom
+    const UTF8_FILE: [u8; 6] = [0xef, 0xbb, 0xbf, b'A', b'B', b'C'];
+
+    // "ABC" in UTF16 with bom
+    const UTF16LE_FILE: [u8; 8] = [0xff, 0xfe, b'A', 0x00, b'B', 0x00, b'C', 0x00];
+    const UTF16BE_FILE: [u8; 8] = [0xfe, 0xff, 0x00, b'A', 0x00, b'B', 0x00, b'C'];
+
+    edit_file_with_content(&UTF8_FILE).await?;
+    edit_file_with_content(&UTF16LE_FILE).await?;
+    edit_file_with_content(&UTF16BE_FILE).await?;
+
+    Ok(())
+}
+
+async fn edit_file_with_content(file_content: &[u8]) -> anyhow::Result<()> {
+    let mut file = tempfile::NamedTempFile::new()?;
+
+    file.as_file_mut().write_all(&file_content)?;
+
+    helpers::test_key_sequence(
+        &mut helpers::AppBuilder::new().build()?,
+        Some(&format!(":o {}<ret>:x<ret>", file.path().to_string_lossy())),
+        None,
+        true,
+    )
+    .await?;
+
+    file.rewind()?;
+    let mut new_file_content: Vec<u8> = Vec::new();
+    file.read_to_end(&mut new_file_content)?;
+
+    assert_eq!(file_content, new_file_content);
+
+    Ok(())
+}

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -133,6 +133,42 @@ impl EncodingBom {
             EncodingBom::NoBom(encoding) => encoding,
         }
     }
+
+    fn for_bom(encoding: &'static Encoding) -> Self {
+        if encoding == encoding::UTF_8 {
+            EncodingBom::UTF8
+        } else if encoding == encoding::UTF_16BE {
+            EncodingBom::UTF16BE
+        } else if encoding == encoding::UTF_16LE {
+            EncodingBom::UTF16LE
+        } else {
+            EncodingBom::NoBom(encoding)
+        }
+    }
+
+    fn write_bom(&self, buf: &mut [u8; BUF_SIZE]) -> usize {
+        match self {
+            EncodingBom::UTF8 => {
+                buf[0] = 0xef;
+                buf[1] = 0xbb;
+                buf[2] = 0xbf;
+                3
+            }
+            EncodingBom::UTF16BE => {
+                buf[0] = 0xfe;
+                buf[1] = 0xff;
+                2
+            }
+
+            EncodingBom::UTF16LE => {
+                buf[0] = 0xef;
+                buf[1] = 0xff;
+                2
+            }
+
+            EncodingBom::NoBom(_) => 0,
+        }
+    }
 }
 
 impl From<&'static Encoding> for EncodingBom {
@@ -333,14 +369,17 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
     let (encoding, mut decoder, mut slice, mut is_empty) = {
         let read = reader.read(&mut buf)?;
         let is_empty = read == 0;
-        let encoding: EncodingBom = match encoding {
-            Some(encoding) => encoding,
-            None => {
+        let encoding = encoding
+            .or_else(|| {
+                encoding::Encoding::for_bom(&buf)
+                    .map(|(encoding, _)| EncodingBom::for_bom(encoding))
+            })
+            .unwrap_or_else(|| {
                 let mut encoding_detector = chardetng::EncodingDetector::new();
                 encoding_detector.feed(&buf, is_empty);
                 encoding_detector.guess(None, true).into()
-            }
-        };
+            });
+
         let decoder = encoding.encoding().new_decoder();
 
         // If the amount of bytes read from the reader is less than
@@ -440,7 +479,8 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
         .chain(std::iter::once(""));
     let mut buf = [0u8; BUF_SIZE];
     let mut encoder = encoding.encoding().new_encoder();
-    let mut total_written = 0usize;
+    let mut total_written = encoding.write_bom(&mut buf);
+
     for chunk in iter {
         let is_empty = chunk.is_empty();
         let mut total_read = 0usize;

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -5,6 +5,7 @@ use futures_util::future::BoxFuture;
 use futures_util::FutureExt;
 use helix_core::auto_pairs::AutoPairs;
 use helix_core::doc_formatter::TextFormat;
+use helix_core::encoding::Encoding;
 use helix_core::syntax::Highlight;
 use helix_core::text_annotations::{InlineAnnotation, TextAnnotations};
 use helix_core::Range;
@@ -115,6 +116,37 @@ pub struct SavePoint {
     revert: Mutex<Transaction>,
 }
 
+#[derive(Debug, Clone)]
+pub enum EncodingBom {
+    UTF8,
+    UTF16BE,
+    UTF16LE,
+    NoBom(&'static Encoding),
+}
+
+impl EncodingBom {
+    fn encoding(&self) -> &'static Encoding {
+        match self {
+            EncodingBom::UTF8 => encoding::UTF_8,
+            EncodingBom::UTF16BE => encoding::UTF_16BE,
+            EncodingBom::UTF16LE => encoding::UTF_16LE,
+            EncodingBom::NoBom(encoding) => encoding,
+        }
+    }
+}
+
+impl From<&'static Encoding> for EncodingBom {
+    fn from(value: &'static Encoding) -> Self {
+        EncodingBom::NoBom(value)
+    }
+}
+
+impl Default for EncodingBom {
+    fn default() -> Self {
+        EncodingBom::NoBom(encoding::UTF_8)
+    }
+}
+
 pub struct Document {
     pub(crate) id: DocumentId,
     text: Rope,
@@ -129,7 +161,7 @@ pub struct Document {
     pub inlay_hints_oudated: bool,
 
     path: Option<PathBuf>,
-    encoding: &'static encoding::Encoding,
+    encoding: EncodingBom,
 
     pub restore_cursor: bool,
 
@@ -282,8 +314,8 @@ impl fmt::Debug for DocumentInlayHintsId {
 /// be used to override encoding auto-detection.
 pub fn from_reader<R: std::io::Read + ?Sized>(
     reader: &mut R,
-    encoding: Option<&'static encoding::Encoding>,
-) -> Result<(Rope, &'static encoding::Encoding), Error> {
+    encoding: Option<EncodingBom>,
+) -> Result<(Rope, EncodingBom), Error> {
     // These two buffers are 8192 bytes in size each and are used as
     // intermediaries during the decoding process. Text read into `buf`
     // from `reader` is decoded into `buf_out` as UTF-8. Once either
@@ -301,12 +333,15 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
     let (encoding, mut decoder, mut slice, mut is_empty) = {
         let read = reader.read(&mut buf)?;
         let is_empty = read == 0;
-        let encoding = encoding.unwrap_or_else(|| {
-            let mut encoding_detector = chardetng::EncodingDetector::new();
-            encoding_detector.feed(&buf, is_empty);
-            encoding_detector.guess(None, true)
-        });
-        let decoder = encoding.new_decoder();
+        let encoding: EncodingBom = match encoding {
+            Some(encoding) => encoding,
+            None => {
+                let mut encoding_detector = chardetng::EncodingDetector::new();
+                encoding_detector.feed(&buf, is_empty);
+                encoding_detector.guess(None, true).into()
+            }
+        };
+        let decoder = encoding.encoding().new_decoder();
 
         // If the amount of bytes read from the reader is less than
         // `buf.len()`, it is undesirable to read the bytes afterwards.
@@ -390,7 +425,7 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
 /// replacement characters may appear in the encoded text.
 pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
     writer: &'a mut W,
-    encoding: &'static encoding::Encoding,
+    encoding: EncodingBom,
     rope: &'a Rope,
 ) -> Result<(), Error> {
     // Text inside a `Rope` is stored as non-contiguous blocks of data called
@@ -404,7 +439,7 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
         .filter(|c| !c.is_empty())
         .chain(std::iter::once(""));
     let mut buf = [0u8; BUF_SIZE];
-    let mut encoder = encoding.new_encoder();
+    let mut encoder = encoding.encoding().new_encoder();
     let mut total_written = 0usize;
     for chunk in iter {
         let is_empty = chunk.is_empty();
@@ -463,10 +498,10 @@ use url::Url;
 impl Document {
     pub fn from(
         text: Rope,
-        encoding: Option<&'static encoding::Encoding>,
+        encoding: Option<EncodingBom>,
         config: Arc<dyn DynAccess<Config>>,
     ) -> Self {
-        let encoding = encoding.unwrap_or(encoding::UTF_8);
+        let encoding = encoding.unwrap_or_default();
         let changes = ChangeSet::new(&text);
         let old_state = None;
 
@@ -507,7 +542,7 @@ impl Document {
     /// overwritten with the `encoding` parameter.
     pub fn open(
         path: &Path,
-        encoding: Option<&'static encoding::Encoding>,
+        encoding: Option<EncodingBom>,
         config_loader: Option<Arc<syntax::Loader>>,
         config: Arc<dyn DynAccess<Config>>,
     ) -> Result<Self, Error> {
@@ -517,7 +552,7 @@ impl Document {
                 std::fs::File::open(path).context(format!("unable to open {:?}", path))?;
             from_reader(&mut file, encoding)?
         } else {
-            let encoding = encoding.unwrap_or(encoding::UTF_8);
+            let encoding = encoding.unwrap_or_default();
             (Rope::from(DEFAULT_LINE_ENDING.as_str()), encoding)
         };
 
@@ -572,7 +607,7 @@ impl Document {
                     })?;
                 {
                     let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
-                    to_writer(&mut stdin, encoding::UTF_8, &text)
+                    to_writer(&mut stdin, EncodingBom::default(), &text)
                         .await
                         .map_err(|_| FormatterError::BrokenStdin)?;
                 }
@@ -684,7 +719,7 @@ impl Document {
         let current_rev = self.get_current_revision();
         let doc_id = self.id();
 
-        let encoding = self.encoding;
+        let encoding = self.encoding.clone();
 
         let last_saved_time = self.last_saved_time;
 
@@ -772,7 +807,7 @@ impl Document {
         provider_registry: &DiffProviderRegistry,
         redraw_handle: RedrawHandle,
     ) -> Result<(), Error> {
-        let encoding = &self.encoding;
+        let encoding = self.encoding.clone();
         let path = self
             .path()
             .filter(|path| path.exists())
@@ -806,14 +841,17 @@ impl Document {
 
     /// Sets the [`Document`]'s encoding with the encoding correspondent to `label`.
     pub fn set_encoding(&mut self, label: &str) -> Result<(), Error> {
-        self.encoding = encoding::Encoding::for_label(label.as_bytes())
-            .ok_or_else(|| anyhow!("unknown encoding"))?;
+        let encoding =
+            Encoding::for_label(label.as_bytes()).ok_or_else(|| anyhow!("unknown encoding"))?;
+
+        self.encoding = encoding.into();
+
         Ok(())
     }
 
     /// Returns the [`Document`]'s current encoding.
-    pub fn encoding(&self) -> &'static encoding::Encoding {
-        self.encoding
+    pub fn encoding(&self) -> &'static Encoding {
+        self.encoding.encoding()
     }
 
     pub fn set_path(&mut self, path: Option<&Path>) -> Result<(), std::io::Error> {
@@ -1271,7 +1309,9 @@ impl Document {
 
     /// Intialize/updates the differ for this document with a new base.
     pub fn set_diff_base(&mut self, diff_base: Vec<u8>, redraw_handle: RedrawHandle) {
-        if let Ok((diff_base, _)) = from_reader(&mut diff_base.as_slice(), Some(self.encoding)) {
+        if let Ok((diff_base, _)) =
+            from_reader(&mut diff_base.as_slice(), Some(self.encoding.clone()))
+        {
             if let Some(differ) = &self.diff_handle {
                 differ.update_diff_base(diff_base);
                 return;
@@ -1715,7 +1755,7 @@ mod test {
                 assert!(ref_path.exists());
 
                 let mut file = std::fs::File::open(path).unwrap();
-                let text = from_reader(&mut file, Some(encoding))
+                let text = from_reader(&mut file, Some(encoding.into()))
                     .unwrap()
                     .0
                     .to_string();
@@ -1741,7 +1781,7 @@ mod test {
 
                 let text = Rope::from_str(&std::fs::read_to_string(path).unwrap());
                 let mut buf: Vec<u8> = Vec::new();
-                helix_lsp::block_on(to_writer(&mut buf, encoding, &text)).unwrap();
+                helix_lsp::block_on(to_writer(&mut buf, encoding.into(), &text)).unwrap();
 
                 let expectation = std::fs::read(ref_path).unwrap();
                 assert_eq!(buf, expectation);

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -116,78 +116,6 @@ pub struct SavePoint {
     revert: Mutex<Transaction>,
 }
 
-// An enum representing the encoding and if it contain a byte order mark
-#[derive(Debug, Clone)]
-pub enum EncodingBom {
-    UTF8,
-    UTF16BE,
-    UTF16LE,
-    NoBom(&'static Encoding),
-}
-
-impl EncodingBom {
-    fn encoding(&self) -> &'static Encoding {
-        match self {
-            EncodingBom::UTF8 => encoding::UTF_8,
-            EncodingBom::UTF16BE => encoding::UTF_16BE,
-            EncodingBom::UTF16LE => encoding::UTF_16LE,
-            EncodingBom::NoBom(encoding) => encoding,
-        }
-    }
-
-    // Assume that encoding have a bom, used to convert
-    // `encoding_rs` for_bom result to this enum
-    fn for_bom(encoding: &'static Encoding) -> Self {
-        if encoding == encoding::UTF_8 {
-            EncodingBom::UTF8
-        } else if encoding == encoding::UTF_16BE {
-            EncodingBom::UTF16BE
-        } else if encoding == encoding::UTF_16LE {
-            EncodingBom::UTF16LE
-        } else {
-            EncodingBom::NoBom(encoding)
-        }
-    }
-
-    // Write in buf the Bom if applicable
-    // return the number of bytes written
-    fn write_bom(&self, buf: &mut [u8; BUF_SIZE]) -> usize {
-        match self {
-            EncodingBom::UTF8 => {
-                buf[0] = 0xef;
-                buf[1] = 0xbb;
-                buf[2] = 0xbf;
-                3
-            }
-            EncodingBom::UTF16BE => {
-                buf[0] = 0xfe;
-                buf[1] = 0xff;
-                2
-            }
-
-            EncodingBom::UTF16LE => {
-                buf[1] = 0xff;
-                buf[0] = 0xfe;
-                2
-            }
-
-            EncodingBom::NoBom(_) => 0,
-        }
-    }
-}
-
-impl From<&'static Encoding> for EncodingBom {
-    fn from(value: &'static Encoding) -> Self {
-        EncodingBom::NoBom(value)
-    }
-}
-
-impl Default for EncodingBom {
-    fn default() -> Self {
-        EncodingBom::NoBom(encoding::UTF_8)
-    }
-}
-
 pub struct Document {
     pub(crate) id: DocumentId,
     text: Rope,
@@ -202,7 +130,8 @@ pub struct Document {
     pub inlay_hints_oudated: bool,
 
     path: Option<PathBuf>,
-    encoding: EncodingBom,
+    encoding: &'static encoding::Encoding,
+    has_bom: bool,
 
     pub restore_cursor: bool,
 
@@ -351,12 +280,12 @@ impl fmt::Debug for DocumentInlayHintsId {
 // its sibling function, `to_writer()`.
 //
 /// Decodes a stream of bytes into UTF-8, returning a `Rope` and the
-/// encoding it was decoded as. The optional `encoding` parameter can
-/// be used to override encoding auto-detection.
+/// encoding it was decoded as with BOM information. The optional `encoding`
+/// parameter can be used to override encoding auto-detection.
 pub fn from_reader<R: std::io::Read + ?Sized>(
     reader: &mut R,
-    encoding: Option<EncodingBom>,
-) -> Result<(Rope, EncodingBom), Error> {
+    encoding: Option<&'static Encoding>,
+) -> Result<(Rope, &'static Encoding, bool), Error> {
     // These two buffers are 8192 bytes in size each and are used as
     // intermediaries during the decoding process. Text read into `buf`
     // from `reader` is decoded into `buf_out` as UTF-8. Once either
@@ -372,26 +301,26 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
     // As a manual override to this auto-detection is possible, the
     // same data is read into `buf` to ensure symmetry in the upcoming
     // loop.
-    let (encoding, mut decoder, mut slice, mut is_empty) = {
+    let (encoding, has_bom, mut decoder, mut slice, mut is_empty) = {
         let read = reader.read(&mut buf)?;
         let is_empty = read == 0;
-        let encoding = encoding
+        let (encoding, has_bom) = encoding
+            .map(|encoding| (encoding, false))
             .or_else(|| {
-                encoding::Encoding::for_bom(&buf)
-                    .map(|(encoding, _)| EncodingBom::for_bom(encoding))
+                encoding::Encoding::for_bom(&buf).map(|(encoding, _bom_size)| (encoding, true))
             })
             .unwrap_or_else(|| {
                 let mut encoding_detector = chardetng::EncodingDetector::new();
                 encoding_detector.feed(&buf, is_empty);
-                encoding_detector.guess(None, true).into()
+                (encoding_detector.guess(None, true), false)
             });
 
-        let decoder = encoding.encoding().new_decoder();
+        let decoder = encoding.new_decoder();
 
         // If the amount of bytes read from the reader is less than
         // `buf.len()`, it is undesirable to read the bytes afterwards.
         let slice = &buf[..read];
-        (encoding, decoder, slice, is_empty)
+        (encoding, has_bom, decoder, slice, is_empty)
     };
 
     // `RopeBuilder::append()` expects a `&str`, so this is the "real"
@@ -459,7 +388,7 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
         is_empty = read == 0;
     }
     let rope = builder.finish();
-    Ok((rope, encoding))
+    Ok((rope, encoding, has_bom))
 }
 
 // The documentation and implementation of this function should be up-to-date with
@@ -470,7 +399,7 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
 /// replacement characters may appear in the encoded text.
 pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
     writer: &'a mut W,
-    encoding: EncodingBom,
+    encoding_with_bom_info: (&'static Encoding, bool),
     rope: &'a Rope,
 ) -> Result<(), Error> {
     // Text inside a `Rope` is stored as non-contiguous blocks of data called
@@ -479,6 +408,7 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
     // determined by filtering the iterator to remove all empty chunks and then
     // appending an empty chunk to it. This is valuable for detecting when all
     // chunks in the `Rope` have been iterated over in the subsequent loop.
+    let (encoding, has_bom) = encoding_with_bom_info;
 
     let iter = rope
         .chunks()
@@ -486,23 +416,32 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
         .chain(std::iter::once(""));
     let mut buf = [0u8; BUF_SIZE];
 
-    let mut total_written = encoding.write_bom(&mut buf);
-
-    let encoding = encoding.encoding();
+    let mut total_written = 0usize;
     let mut encoder = encoding.new_encoder();
 
     const UTF_16_BOM: u16 = 0xfeff;
     if encoding == encoding::UTF_16BE {
-        writer.write_u16(UTF_16_BOM).await?;
+        if has_bom {
+            writer.write_u16(UTF_16_BOM).await?;
+        }
         for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
             writer.write_u16(ch).await?;
         }
     } else if encoding == encoding::UTF_16LE {
-        writer.write_u16_le(UTF_16_BOM).await?;
+        if has_bom {
+            writer.write_u16_le(UTF_16_BOM).await?;
+        }
         for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
             writer.write_u16_le(ch).await?;
         }
     } else {
+        if encoding == encoding::UTF_8 && has_bom {
+            buf[0] = 0xef;
+            buf[1] = 0xbb;
+            buf[2] = 0xbf;
+            total_written = 3;
+        }
+
         for chunk in iter {
             let is_empty = chunk.is_empty();
             let mut total_read = 0usize;
@@ -565,10 +504,10 @@ use url::Url;
 impl Document {
     pub fn from(
         text: Rope,
-        encoding: Option<EncodingBom>,
+        encoding_with_bom_info: Option<(&'static Encoding, bool)>,
         config: Arc<dyn DynAccess<Config>>,
     ) -> Self {
-        let encoding = encoding.unwrap_or_default();
+        let (encoding, has_bom) = encoding_with_bom_info.unwrap_or((encoding::UTF_8, false));
         let changes = ChangeSet::new(&text);
         let old_state = None;
 
@@ -576,6 +515,7 @@ impl Document {
             id: DocumentId::default(),
             path: None,
             encoding,
+            has_bom,
             text,
             selections: HashMap::default(),
             inlay_hints: HashMap::default(),
@@ -609,21 +549,21 @@ impl Document {
     /// overwritten with the `encoding` parameter.
     pub fn open(
         path: &Path,
-        encoding: Option<EncodingBom>,
+        encoding: Option<&'static Encoding>,
         config_loader: Option<Arc<syntax::Loader>>,
         config: Arc<dyn DynAccess<Config>>,
     ) -> Result<Self, Error> {
         // Open the file if it exists, otherwise assume it is a new file (and thus empty).
-        let (rope, encoding) = if path.exists() {
+        let (rope, encoding, has_bom) = if path.exists() {
             let mut file =
                 std::fs::File::open(path).context(format!("unable to open {:?}", path))?;
             from_reader(&mut file, encoding)?
         } else {
-            let encoding = encoding.unwrap_or_default();
-            (Rope::from(DEFAULT_LINE_ENDING.as_str()), encoding)
+            let encoding = encoding.unwrap_or(encoding::UTF_8);
+            (Rope::from(DEFAULT_LINE_ENDING.as_str()), encoding, false)
         };
 
-        let mut doc = Self::from(rope, Some(encoding), config);
+        let mut doc = Self::from(rope, Some((encoding, has_bom)), config);
 
         // set the path and try detecting the language
         doc.set_path(Some(path))?;
@@ -674,7 +614,7 @@ impl Document {
                     })?;
                 {
                     let mut stdin = process.stdin.take().ok_or(FormatterError::BrokenStdin)?;
-                    to_writer(&mut stdin, EncodingBom::default(), &text)
+                    to_writer(&mut stdin, (encoding::UTF_8, false), &text)
                         .await
                         .map_err(|_| FormatterError::BrokenStdin)?;
                 }
@@ -786,8 +726,7 @@ impl Document {
         let current_rev = self.get_current_revision();
         let doc_id = self.id();
 
-        let encoding = self.encoding.clone();
-
+        let encoding_with_bom_info = (self.encoding, self.has_bom);
         let last_saved_time = self.last_saved_time;
 
         // We encode the file according to the `Document`'s encoding.
@@ -816,7 +755,7 @@ impl Document {
             }
 
             let mut file = File::create(&path).await?;
-            to_writer(&mut file, encoding, &text).await?;
+            to_writer(&mut file, encoding_with_bom_info, &text).await?;
 
             let event = DocumentSavedEvent {
                 revision: current_rev,
@@ -874,7 +813,7 @@ impl Document {
         provider_registry: &DiffProviderRegistry,
         redraw_handle: RedrawHandle,
     ) -> Result<(), Error> {
-        let encoding = self.encoding.clone();
+        let encoding = self.encoding;
         let path = self
             .path()
             .filter(|path| path.exists())
@@ -911,14 +850,14 @@ impl Document {
         let encoding =
             Encoding::for_label(label.as_bytes()).ok_or_else(|| anyhow!("unknown encoding"))?;
 
-        self.encoding = encoding.into();
+        self.encoding = encoding;
 
         Ok(())
     }
 
     /// Returns the [`Document`]'s current encoding.
     pub fn encoding(&self) -> &'static Encoding {
-        self.encoding.encoding()
+        self.encoding
     }
 
     pub fn set_path(&mut self, path: Option<&Path>) -> Result<(), std::io::Error> {
@@ -1376,9 +1315,7 @@ impl Document {
 
     /// Intialize/updates the differ for this document with a new base.
     pub fn set_diff_base(&mut self, diff_base: Vec<u8>, redraw_handle: RedrawHandle) {
-        if let Ok((diff_base, _)) =
-            from_reader(&mut diff_base.as_slice(), Some(self.encoding.clone()))
-        {
+        if let Ok((diff_base, ..)) = from_reader(&mut diff_base.as_slice(), Some(self.encoding)) {
             if let Some(differ) = &self.diff_handle {
                 differ.update_diff_base(diff_base);
                 return;
@@ -1848,7 +1785,7 @@ mod test {
 
                 let text = Rope::from_str(&std::fs::read_to_string(path).unwrap());
                 let mut buf: Vec<u8> = Vec::new();
-                helix_lsp::block_on(to_writer(&mut buf, encoding.into(), &text)).unwrap();
+                helix_lsp::block_on(to_writer(&mut buf, (encoding, false), &text)).unwrap();
 
                 let expectation = std::fs::read(ref_path).unwrap();
                 assert_eq!(buf, expectation);

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -166,8 +166,8 @@ impl EncodingBom {
             }
 
             EncodingBom::UTF16LE => {
-                buf[0] = 0xef;
                 buf[1] = 0xff;
+                buf[0] = 0xfe;
                 2
             }
 
@@ -479,54 +479,75 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
     // determined by filtering the iterator to remove all empty chunks and then
     // appending an empty chunk to it. This is valuable for detecting when all
     // chunks in the `Rope` have been iterated over in the subsequent loop.
+
     let iter = rope
         .chunks()
         .filter(|c| !c.is_empty())
         .chain(std::iter::once(""));
     let mut buf = [0u8; BUF_SIZE];
-    let mut encoder = encoding.encoding().new_encoder();
+
     let mut total_written = encoding.write_bom(&mut buf);
 
-    for chunk in iter {
-        let is_empty = chunk.is_empty();
-        let mut total_read = 0usize;
+    let encoding = encoding.encoding();
+    let mut encoder = encoding.new_encoder();
 
-        // An inner loop is necessary as it is possible that the input buffer
-        // may not be completely encoded on the first `encode_from_utf8()` call
-        // which would happen in cases where the output buffer is filled to
-        // capacity.
-        loop {
-            let (result, read, written, ..) =
-                encoder.encode_from_utf8(&chunk[total_read..], &mut buf[total_written..], is_empty);
+    const UTF_16_BOM: u16 = 0xfeff;
+    if encoding == encoding::UTF_16BE {
+        writer.write_u16(UTF_16_BOM).await?;
+        for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
+            writer.write_u16(ch).await?;
+        }
+    } else if encoding == encoding::UTF_16LE {
+        writer.write_u16_le(UTF_16_BOM).await?;
+        for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
+            writer.write_u16_le(ch).await?;
+        }
+    } else {
+        for chunk in iter {
+            let is_empty = chunk.is_empty();
+            let mut total_read = 0usize;
 
-            // These variables act as the read and write cursors of `chunk` and `buf` respectively.
-            // They are necessary in case the output buffer fills before encoding of the entire input
-            // loop is complete. Otherwise, the loop would endlessly iterate over the same `chunk` and
-            // the data inside the output buffer would be overwritten.
-            total_read += read;
-            total_written += written;
-            match result {
-                encoding::CoderResult::InputEmpty => {
-                    debug_assert_eq!(chunk.len(), total_read);
-                    debug_assert!(buf.len() >= total_written);
-                    break;
-                }
-                encoding::CoderResult::OutputFull => {
-                    debug_assert!(chunk.len() > total_read);
-                    writer.write_all(&buf[..total_written]).await?;
-                    total_written = 0;
+            // An inner loop is necessary as it is possible that the input buffer
+            // may not be completely encoded on the first `encode_from_utf8()` call
+            // which would happen in cases where the output buffer is filled to
+            // capacity.
+            loop {
+                let (result, read, written, ..) = encoder.encode_from_utf8(
+                    &chunk[total_read..],
+                    &mut buf[total_written..],
+                    is_empty,
+                );
+
+                // These variables act as the read and write cursors of `chunk` and `buf` respectively.
+                // They are necessary in case the output buffer fills before encoding of the entire input
+                // loop is complete. Otherwise, the loop would endlessly iterate over the same `chunk` and
+                // the data inside the output buffer would be overwritten.
+                total_read += read;
+                total_written += written;
+                match result {
+                    encoding::CoderResult::InputEmpty => {
+                        debug_assert_eq!(chunk.len(), total_read);
+                        debug_assert!(buf.len() >= total_written);
+                        break;
+                    }
+                    encoding::CoderResult::OutputFull => {
+                        debug_assert!(chunk.len() > total_read);
+                        writer.write_all(&buf[..total_written]).await?;
+                        total_written = 0;
+                    }
                 }
             }
-        }
 
-        // Once the end of the iterator is reached, the output buffer is
-        // flushed and the outer loop terminates.
-        if is_empty {
-            writer.write_all(&buf[..total_written]).await?;
-            writer.flush().await?;
-            break;
+            // Once the end of the iterator is reached, the output buffer is
+            // flushed and the outer loop terminates.
+            if is_empty {
+                writer.write_all(&buf[..total_written]).await?;
+                writer.flush().await?;
+                break;
+            }
         }
     }
+
     Ok(())
 }
 

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -276,6 +276,94 @@ impl fmt::Debug for DocumentInlayHintsId {
     }
 }
 
+enum Encoder {
+    Utf16Be,
+    Utf16Le,
+    EncodingRs(encoding::Encoder),
+}
+
+impl Encoder {
+    fn from_encoding(encoding: &'static encoding::Encoding) -> Self {
+        if encoding == encoding::UTF_16BE {
+            Self::Utf16Be
+        } else if encoding == encoding::UTF_16LE {
+            Self::Utf16Le
+        } else {
+            Self::EncodingRs(encoding.new_encoder())
+        }
+    }
+
+    fn encode_from_utf8(
+        &mut self,
+        src: &str,
+        dst: &mut [u8],
+        is_empty: bool,
+    ) -> (encoding::CoderResult, usize, usize) {
+        if src.is_empty() {
+            return (encoding::CoderResult::InputEmpty, 0, 0);
+        }
+        let mut write_to_buf = |convert: fn(u16) -> [u8; 2]| {
+            let to_write = src.char_indices().map(|(indice, char)| {
+                let mut encoded: [u16; 2] = [0, 0];
+                (
+                    indice,
+                    char.encode_utf16(&mut encoded)
+                        .iter_mut()
+                        .flat_map(|char| convert(*char))
+                        .collect::<Vec<u8>>(),
+                )
+            });
+
+            let mut total_written = 0usize;
+
+            for (indice, utf16_bytes) in to_write {
+                let character_size = utf16_bytes.len();
+
+                if dst.len() <= (total_written + character_size) {
+                    return (encoding::CoderResult::OutputFull, indice, total_written);
+                }
+
+                for character in utf16_bytes {
+                    dst[total_written] = character;
+                    total_written += 1;
+                }
+            }
+
+            (encoding::CoderResult::InputEmpty, src.len(), total_written)
+        };
+
+        match self {
+            Self::Utf16Be => write_to_buf(u16::to_be_bytes),
+            Self::Utf16Le => write_to_buf(u16::to_le_bytes),
+            Self::EncodingRs(encoder) => {
+                let (code_result, read, written, ..) = encoder.encode_from_utf8(src, dst, is_empty);
+
+                (code_result, read, written)
+            }
+        }
+    }
+}
+
+// Apply BOM if encoding permit it, return the number of bytes written at the start of buf
+fn apply_bom(encoding: &'static encoding::Encoding, buf: &mut [u8; BUF_SIZE]) -> usize {
+    if encoding == encoding::UTF_8 {
+        buf[0] = 0xef;
+        buf[1] = 0xbb;
+        buf[2] = 0xbf;
+        3
+    } else if encoding == encoding::UTF_16BE {
+        buf[0] = 0xfe;
+        buf[1] = 0xff;
+        2
+    } else if encoding == encoding::UTF_16LE {
+        buf[0] = 0xff;
+        buf[1] = 0xfe;
+        2
+    } else {
+        0
+    }
+}
+
 // The documentation and implementation of this function should be up-to-date with
 // its sibling function, `to_writer()`.
 //
@@ -416,74 +504,52 @@ pub async fn to_writer<'a, W: tokio::io::AsyncWriteExt + Unpin + ?Sized>(
         .chain(std::iter::once(""));
     let mut buf = [0u8; BUF_SIZE];
 
-    let mut total_written = 0usize;
-    let mut encoder = encoding.new_encoder();
-
-    const UTF_16_BOM: u16 = 0xfeff;
-    if encoding == encoding::UTF_16BE {
-        if has_bom {
-            writer.write_u16(UTF_16_BOM).await?;
-        }
-        for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
-            writer.write_u16(ch).await?;
-        }
-    } else if encoding == encoding::UTF_16LE {
-        if has_bom {
-            writer.write_u16_le(UTF_16_BOM).await?;
-        }
-        for ch in rope.chunks().flat_map(|chunk| chunk.encode_utf16()) {
-            writer.write_u16_le(ch).await?;
-        }
+    let mut total_written = if has_bom {
+        apply_bom(encoding, &mut buf)
     } else {
-        if encoding == encoding::UTF_8 && has_bom {
-            buf[0] = 0xef;
-            buf[1] = 0xbb;
-            buf[2] = 0xbf;
-            total_written = 3;
-        }
+        0
+    };
 
-        for chunk in iter {
-            let is_empty = chunk.is_empty();
-            let mut total_read = 0usize;
+    let mut encoder = Encoder::from_encoding(encoding);
 
-            // An inner loop is necessary as it is possible that the input buffer
-            // may not be completely encoded on the first `encode_from_utf8()` call
-            // which would happen in cases where the output buffer is filled to
-            // capacity.
-            loop {
-                let (result, read, written, ..) = encoder.encode_from_utf8(
-                    &chunk[total_read..],
-                    &mut buf[total_written..],
-                    is_empty,
-                );
+    for chunk in iter {
+        let is_empty = chunk.is_empty();
+        let mut total_read = 0usize;
 
-                // These variables act as the read and write cursors of `chunk` and `buf` respectively.
-                // They are necessary in case the output buffer fills before encoding of the entire input
-                // loop is complete. Otherwise, the loop would endlessly iterate over the same `chunk` and
-                // the data inside the output buffer would be overwritten.
-                total_read += read;
-                total_written += written;
-                match result {
-                    encoding::CoderResult::InputEmpty => {
-                        debug_assert_eq!(chunk.len(), total_read);
-                        debug_assert!(buf.len() >= total_written);
-                        break;
-                    }
-                    encoding::CoderResult::OutputFull => {
-                        debug_assert!(chunk.len() > total_read);
-                        writer.write_all(&buf[..total_written]).await?;
-                        total_written = 0;
-                    }
+        // An inner loop is necessary as it is possible that the input buffer
+        // may not be completely encoded on the first `encode_from_utf8()` call
+        // which would happen in cases where the output buffer is filled to
+        // capacity.
+        loop {
+            let (result, read, written, ..) =
+                encoder.encode_from_utf8(&chunk[total_read..], &mut buf[total_written..], is_empty);
+
+            // These variables act as the read and write cursors of `chunk` and `buf` respectively.
+            // They are necessary in case the output buffer fills before encoding of the entire input
+            // loop is complete. Otherwise, the loop would endlessly iterate over the same `chunk` and
+            // the data inside the output buffer would be overwritten.
+            total_read += read;
+            total_written += written;
+            match result {
+                encoding::CoderResult::InputEmpty => {
+                    debug_assert_eq!(chunk.len(), total_read);
+                    debug_assert!(buf.len() >= total_written);
+                    break;
+                }
+                encoding::CoderResult::OutputFull => {
+                    debug_assert!(chunk.len() > total_read);
+                    writer.write_all(&buf[..total_written]).await?;
+                    total_written = 0;
                 }
             }
+        }
 
-            // Once the end of the iterator is reached, the output buffer is
-            // flushed and the outer loop terminates.
-            if is_empty {
-                writer.write_all(&buf[..total_written]).await?;
-                writer.flush().await?;
-                break;
-            }
+        // Once the end of the iterator is reached, the output buffer is
+        // flushed and the outer loop terminates.
+        if is_empty {
+            writer.write_all(&buf[..total_written]).await?;
+            writer.flush().await?;
+            break;
         }
     }
 

--- a/helix-view/src/document.rs
+++ b/helix-view/src/document.rs
@@ -116,6 +116,7 @@ pub struct SavePoint {
     revert: Mutex<Transaction>,
 }
 
+// An enum representing the encoding and if it contain a byte order mark
 #[derive(Debug, Clone)]
 pub enum EncodingBom {
     UTF8,
@@ -134,6 +135,8 @@ impl EncodingBom {
         }
     }
 
+    // Assume that encoding have a bom, used to convert
+    // `encoding_rs` for_bom result to this enum
     fn for_bom(encoding: &'static Encoding) -> Self {
         if encoding == encoding::UTF_8 {
             EncodingBom::UTF8
@@ -146,6 +149,8 @@ impl EncodingBom {
         }
     }
 
+    // Write in buf the Bom if applicable
+    // return the number of bytes written
     fn write_bom(&self, buf: &mut [u8; BUF_SIZE]) -> usize {
         match self {
             EncodingBom::UTF8 => {
@@ -361,8 +366,9 @@ pub fn from_reader<R: std::io::Read + ?Sized>(
     let mut buf_out = [0u8; BUF_SIZE];
     let mut builder = RopeBuilder::new();
 
-    // By default, the encoding of the text is auto-detected via the
-    // `chardetng` crate which requires sample data from the reader.
+    // By default, the encoding of the text is auto-detected by
+    // `encoding_rs` for_bom, and if it fails, from `chardetng`
+    // crate which requires sample data from the reader.
     // As a manual override to this auto-detection is possible, the
     // same data is read into `buf` to ensure symmetry in the upcoming
     // loop.

--- a/helix-view/src/editor.rs
+++ b/helix-view/src/editor.rs
@@ -1,7 +1,7 @@
 use crate::{
     align_view,
     clipboard::{get_clipboard_provider, ClipboardProvider},
-    document::{DocumentSavedEventFuture, DocumentSavedEventResult, EncodingBom, Mode},
+    document::{DocumentSavedEventFuture, DocumentSavedEventResult, Mode},
     graphics::{CursorKind, Rect},
     info::Info,
     input::KeyEvent,

--- a/helix-view/src/editor.rs
+++ b/helix-view/src/editor.rs
@@ -1299,10 +1299,10 @@ impl Editor {
     }
 
     pub fn new_file_from_stdin(&mut self, action: Action) -> Result<DocumentId, Error> {
-        let (rope, encoding) = crate::document::from_reader(&mut stdin(), None)?;
+        let (rope, encoding, has_bom) = crate::document::from_reader(&mut stdin(), None)?;
         Ok(self.new_file_from_document(
             action,
-            Document::from(rope, Some(encoding), self.config.clone()),
+            Document::from(rope, Some((encoding, has_bom)), self.config.clone()),
         ))
     }
 

--- a/helix-view/src/editor.rs
+++ b/helix-view/src/editor.rs
@@ -1,7 +1,7 @@
 use crate::{
     align_view,
     clipboard::{get_clipboard_provider, ClipboardProvider},
-    document::{DocumentSavedEventFuture, DocumentSavedEventResult, Mode},
+    document::{DocumentSavedEventFuture, DocumentSavedEventResult, EncodingBom, Mode},
     graphics::{CursorKind, Rect},
     info::Info,
     input::KeyEvent,

